models/issue.go

```diff
@@ -1237,6 +1237,24 @@ func CountIssuesByRepo(opts *IssuesOptions) (map[int64]int64, error) {
 	return countMap, nil
 }
 
+// GetRepoIDsForIssuesOptions find all repo ids for the given options
+func GetRepoIDsForIssuesOptions(opts *IssuesOptions) ([]int64, error) {
+	repoIDs := make([]int64, 0, 5)
+	sess := x.NewSession()
+	defer sess.Close()
+
+	opts.setupSession(sess)
+
+	if err := sess.GroupBy("issue.repo_id").
+		Distinct("issue.repo_id").
+		Table("issue").
+		Find(&repoIDs); err != nil {
```

Comment:
This seems to get a list of all of the repositories in the system that contain issues or PRs. I think we should filter by visibility to get a shorter list.

IMHO it would be much faster to bring repository ids from the `repository` table (even if they don't have any issues).

(Also, I might be wrong, but I think you can get rid of the `GroupBy` option in this particular case; it could be significantly faster in some scenarios if xorm allows it)

Reply:
Are you aware of models.accessibleRepositoryCondition

Reply:
@guillep2k I see the method doesn't look very clear here, but the call on line 1246: "opts.setupSession(sess)" adds other conditions to the query. The caller (Issues method in routers/user/home.go") will have set appropriate options in the IssuesOptions struct especially based on the filter mode i.e. "In your repositories", "Assigned to you", etc. The idea is to get the list of all the repo ids we need to search for the keyword in, which might be a different list depending on the filter mode. So no the intent isn't to get a list of all repos with issues/PRs.

Reply:
> Are you aware of models.accessibleRepositoryCondition

No, I wasn't. Do you think I should add that to the query? I saw a couple examples from a code search

Reply:
>
> @guillep2k I see the method doesn't look very clear here, but the call on line 1246: "opts.setupSession(sess)" adds other conditions to the query.

Oh, I see. Have you checked if visibility (i.e. permissions) is already taken care of at the time of call? Otherwise you'll have to use `models.accessibleRepositoryCondition` as pointed out by @zeripath

Reply:
@guillep2k yes and no I suppose...the filter mode from the issues page will ensure that a join in the query will only return the appropriate repositories, however, there's no explicit check for access afterwards

Reply:
To be on the safe side I added the "accessibleRepositoryCondition" as suggested by @zeripath

Reply:
I think that instead of:
```
JOIN (SELECT DISTINCT repo_id WHERE ... GROUP BY repo_id)
```
(which by the way is redundant) we should use instead
```
AND repo_id IN (SELECT repo_id FROM ....)
```
The difference is subtle but important: in the first case the database needs to find all the issues in all repositories that match the condition, group them and then perform a `JOIN`. In the second case, most databases will realize that a particular repo_id is already on the `IN` list and skip all the other issues from that repo to save work and memory. Furthermore, the database can be free to choose the other part of the join to drive the query depending on the number of rows involved.

Reply:
I removed the GroupBy("issue.repo_id"). line as it was unnecessary. Thanks for pointing this out. I hope this gets the query to a place you think is good, but let me know if there's still a concern